src/TabulatedCooling.hpp

```diff
@@ -0,0 +1,316 @@
+{
+	// interpolate cooling rates from Cloudy tables
+	const Real rhoH = rho * cloudy_H_mass_fraction; // mass density of H species
+	const Real nH = rhoH / (C::m_p + C::m_e);
+	const Real log_nH = std::log10(nH);
+	const Real log_T = std::log10(T);
+
+	const double logCool = interpolate2d(log_nH, log_T, tables.log_nH, tables.log_Tgas, tables.cool);
+
+	const double logHeat = interpolate2d(log_nH, log_T, tables.log_nH, tables.log_Tgas, tables.heat);
+
+	const double netLambda = FastMath::pow10(logHeat) - FastMath::pow10(logCool);
+
+	// multiply by the square of H mass density (**NOT number density**)
+	const double Edot = (rhoH * rhoH) * netLambda;
+
+	return Edot;
+}
+
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto ComputeEgasFromTgas(double rho, double Tgas, double gamma, cloudyGpuConstTables const &tables) -> Real
+{
+	// convert Egas (internal gas energy) to temperature
+	const Real rhoH = rho * cloudy_H_mass_fraction;
+	const Real nH = rhoH / (C::m_p + C::m_e);
+
+	// compute mu from mu(T) table
+	const Real mu = interpolate2d(std::log10(nH), std::log10(Tgas), tables.log_nH, tables.log_Tgas, tables.meanMolWeight);
+
+	// compute thermal gas energy
+	const Real n = rho / ((C::m_p + C::m_e) * mu);
+	const Real Pgas = n * C::k_B * Tgas;
+	const Real Egas = Pgas / (gamma - 1.);
+	return Egas;
+}
+
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto ComputeTgasFromEgas(double rho, double Egas, double gamma, cloudyGpuConstTables const &tables) -> Real
+{
+	// convert Egas (internal gas energy) to temperature
+
+	// check whether temperature is out-of-bounds
+	const Real Tmin_table = 10.;
+	const Real Tmax_table = 1.0e9;
+	const Real Eint_min = ComputeEgasFromTgas(rho, Tmin_table, gamma, tables);
+	const Real Eint_max = ComputeEgasFromTgas(rho, Tmax_table, gamma, tables);
+
+	if (Egas <= Eint_min) {
+		return Tmin_table;
+	}
+	if (Egas >= Eint_max) {
+		return Tmax_table;
+	}
+
+	// solve for temperature given Eint (with fixed adiabatic index gamma)
+	const Real rhoH = rho * cloudy_H_mass_fraction;
+	const Real nH = rhoH / (C::m_p + C::m_e);
+	const Real log_nH = std::log10(nH);
+
+	// mean molecular weight (in Grackle tables) is defined w/r/t
+	// (C::m_p + C::m_e)
+	const Real C = (gamma - 1.) * Egas / (C::k_B * (rho / (C::m_p + C::m_e)));
+
+	// solve for mu(T)*C == T.
+	// (Grackle does this with a fixed-point iteration. We use a more robust
+	// method, similar to Brent's method, the TOMS748 method.)
+	const Real reltol = 1.0e-5;
+	const int maxIterLimit = 100;
+	int maxIter = maxIterLimit;
+
+	auto f = [log_nH, C, tables](const Real &T) noexcept {
+		// compute new mu from mu(log10 T) table
+		Real const log_T = clamp(std::log10(T), 1., 9.);
+		Real const mu = interpolate2d(log_nH, log_T, tables.log_nH, tables.log_Tgas, tables.meanMolWeight);
+		Real const fun = C * mu - T;
+		return fun;
+	};
+
+	// compute temperature bounds using physics
+	const Real mu_min = 0.60; // assuming fully ionized (mu ~ 0.6)
```

Comment:
I worry that these bounds on mu are not precise, and that you may get outside them if you consider alternative compositions, for example if you vary the He mass fraction between primordial and Milky Way values. I suggest writing more precise bounds in terms of the element mass fractions. My proposed bounds are as follows: first, let X, Y, and Z be the H, He, and metal mass fractions, respectively, and treat metals as having mean atomic weight of A and a mean atomic number A/2 (reasonable if you are dominated by alpha elements). In this case the number of H, He, and metal atoms per unit mass is X / m_H, Y / 4 m_H, and Z / A m_H, and for full ionization the number of free electrons per unit mass is X / m_H + Y / 2 m_H + Z / (2 A m_H). The number of free particles per unit mass for full ionization is therefore 2 X / m_H + 3 Y / 4 m_H + 3 Z / 2 A m_H, and since mu = m_H / free particles per unit mass, we have mu_min = 1 / (2 X + 3 Y / 4 + 3 Z / 2 A); a reasonable choice would be A = 12 (carbon) or A = 16 (oxygen). Conversely, if the H is all molecular and there is no ionization, then the number of free particles per unit mass is X / 2 m_H + Y / 4 m_H + Z / A m_H, and mu = mu_max = 1 / (X/2 + Y/4 + Z/A).

Reply:
Great. And, just as a note for the future, while the expressions here relating temperature, internal energy, and pressure are approximately right given that we are making the approximation of fixed gamma, in the future when we go to fully self-consistent and variable gamma they will need to be revised.

Reply:
Actually, I can just set this by precomputing the minimum and maximum values of the mmw array when the table is read in.

Reply:
OK, that's fine.